play - validate hosts entries

samdoran opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

The `Play.load` method is defining `name` by joining values in `hosts` for the play. This should be done in `Play.get_name()`, allowing the `load` method to focus on loading data. Create a separate method for validating the values in `hosts`.

Fixes #65386
Closes #74054
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`ansible.playbook.play`

samdoran committed:
Add tests for invalid hosts

Most of these fail currently.

samdoran committed:
First rough pass at improving hosts validation

bcoca commented:
i thought we kept the 'untemplated' version of hosts there on purpose to see what was 'stated' to be easy to contrast with what was 'achieved'

samdoran committed:
Validate host inputs

- check for empty values
- check for None
- check for values that are not a sequence and are not strings
- add unit tests

samdoran committed:
Clean up the validation logic a bit more

Add more tests as well.

samdoran committed:
Move play name setting to get_name() and out of load()

samdoran committed:
Change tests to pytest-style tests

samdoran committed:
Use match parameter rather than an assert statement

samdoran committed:
Update error message in integration test

samdoran committed:
Return proper name or empty string, not None

samdoran committed:
Add changelog

samdoran committed:
Increase test coverage

samdoran committed:
Tweak error message and add a test case

samdoran commented:
I have what I think is a pretty complete host validation code. I still think it should probably move outside of `Play.load()` since that is bulk of the complexity in that method and it doesn't really have anything to do with loading the data.

samdoran committed:
Test all the things with coverage

ci_complete
ci_coverage

samdoran commented:
> i thought we kept the 'untemplated' version of hosts there on purpose to see what was 'stated' to be easy to contrast with what was 'achieved'

I'm not sure about that. I do know that not specifying `name` when creating the play object the name will be `''` and `get_name()` becomes the only way to get the actual name.

That's a change in behavior and maybe we _do_ want to make sure a `name` is always passed in when creating a `Play` object?

It seems like we always call `.get_name()` rather than `.name`, so this change might be ok. Would like others to weigh in.

bcoca commented:
This kind of 'prep' work should happen in either preprocess_data or _validate_attributes (base class calls these in 'load' before doing creating the class).

samdoran commented:
> fyi, the normal validate will internally call a _validate_hosts method if this exists, which I believe would be 'final' place to put this

This seems like the best place to put this code. I will try that.

samdoran committed:
Add _validate_hosts() method

By defining this method, it gets called automatically by FieldAttributeBase.validate().

samdoran commented:
Adding a `_validate_hosts()` method worked beautifully.

samdoran committed:
Add a few more tests cases

bcoca commented on lib/ansible/playbook/play.py:
```diff
@@ -97,19 +98,38 @@ def __init__(self):
+        # Only validate 'hosts' if a value was passed in to original data set.
+        if 'hosts' in self._ds:
+            # Check for empty data
+            if value in ([], tuple(), set(), {}, ''):
```
why not `if not value` ?

samdoran replied:
I wanted the error message for `False`, `0`, and `None` to be different. Maybe that's splitting hairs and trying too hard.

samdoran replied:
I will simply this test.

bcoca replied:
i'll wait for someone to name a host or group 'false'  and cry when that ticket is opened

bcoca commented on lib/ansible/playbook/play.py:
```diff
@@ -97,19 +98,38 @@ def __init__(self):
+                for host in value:
+                    if host is None:
+                        raise AnsibleParserError("Hosts list cannot contain values of 'None'. Please check your playbook")
+                    elif not is_string(host):
```
so this is valid
```
- hosts:
    - mailer
    - local1, local2
```
but this is not?
```
- hosts:
    - mailer
    - [local1, local2]
```

bcoca replied:
```- hosts:
    - mailer
    - local1, [local2, localhost]
```
and then again, this is also legal, just does not do what people would expect

bcoca replied:
I would also do `not host` as `''` is both a string and not None, but i would argue still an invalid host

samdoran replied:
Mixing lists and strings with comma separated lists is messy. I'm not sure how far we want to go down the path of validating specific host entry strings.

bcoca replied:
I don't disagree, but that is 'current behavior' and not clearly a buggy use.

bcoca commented:
i would consider a general 'flatten' to allow 'all forms' of lists of lists, but still ensure we have strings in the end

samdoran commented:
I think I'd be ok with flattening lists of lists, but not with transposing and flattening `[host1, 'host2, host3']` into `[host1, host2, host3]`. That's probably too much magic.

samdoran commented:
That this works is hurting my head. 🤦‍♂️ 

```yaml
- hosts:
    - mailer
    - local1, local2
```

samdoran commented:
Currently I made both of these scenarios fail with an error message:

```yaml
- hosts:
    - mailer
    - [local1, local2]
```

```yaml
hosts:
    - mailer
    - local1, [local2, localhost]
```

I am trying out flatting recursive lists, but am concerned about large inventories and how that extra processing could impact performance.

samdoran committed:
Error on lists of lists or a string that is list or dict-like

samdoran committed:
Flatten hosts lists

Also turn comma separated host values into lists.
Add helper funcutions for flattening a list and splitting strings into a list.

samdoran commented:
I implemented changes that do some work in `preprocess_data()` to flatten the hosts list. In my testing, the helper function I created hits recursion errors with 1484 items in the `hosts` list.

With 1483 host entries or fewer, it is slower but we are doing extra work so that's to be expected. I am more concerned about introducing a potential recursion error here.

With the latest changes in this PR:
```
%timeit Play.load({'hosts': [f'host{d}' for d in range(1483)]})
34.7 ms ± 1.43 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Latest `devel`:
```
In [2]: %timeit Play.load({'hosts': [f'host{d}' for d in range(1483)]})
443 µs ± 75.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [3]: %timeit Play.load({'hosts': [f'host{d}' for d in range(1500)]})
369 µs ± 2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [4]: %timeit Play.load({'hosts': [f'host{d}' for d in range(3000)]})
664 µs ± 11.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: %timeit Play.load({'hosts': [f'host{d}' for d in range(10000)]})
1.93 ms ± 23.8 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

samdoran committed:
Cleanup unnecessary comment and breakpoint

bcoca commented:
i'm thinking you don't need to go this far with validation here, since the 'full validation' happens in post_validation functions (after templating and also ensures correct type). 

I think you can just ensure it is 'serializable' at this point (so 'name' works w/o bombing) or force hosts: post validation to go before 'name' post_validation. the difference being name reflecting the 'unprocessed host list' vs 'final host list'.

samdoran commented:
I'll dial this back a bit. It's causing other problems converting to a list too early.

bcoca commented:
the simplest fix might be to catch exception on name construction and just use a filler/default.

s-hertel commented:
I'm in favor of going back to the approach you had before this comment. https://github.com/ansible/ansible/pull/74147#issuecomment-844475894

It was much simpler, keeps parity with the existing code, and we can always add more hosts formats.  I can't think of a good reason why anyone would even want to structure their hosts like
```
hosts:
  - host1
  - host2, [host3, host4]
```

bcoca commented:
note that we hit this problem also when doing --list-X options  https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/playbook.py#L150

samdoran committed:
Revert "Flatten hosts lists"

This reverts commit cd1067c2bcc198b4db2b5067c52e18fe16406983.

# Conflicts:
#	lib/ansible/playbook/play.py

samdoran commented:
I reverted the recursive list flattening. I'm happy with how this works currently. It solves the original issue plus adds more robust testing. It also errors nicely with the `--listX` options.

samdoran committed:
Allow Jinja expressions in the hosts field

samdoran commented:
Seems like I am going to have go further in dialing back some of the validation. Brackets are valid for IP addresses, so failing if they are in a hosts entry will probably cause problems. 😞

samdoran committed:
Do not try to identify things that look like lists within a string

Since '[' and ']' are valid values in IP addresses and could in theory be valid for an inventory
hostname (even though they are not valid characters), considering those values invalid
will probably bring more false positives than actual invalid value detections.

samdoran commented:
I removed the checking for `[` or `]` in a hosts entry. Unfortunately that means we are back to the scenario described above in https://github.com/ansible/ansible/pull/74147#discussion_r634404331.

For now, I think it's probably best to let that go. This change still does fix the original issue and improves the validation without making too many large changes.

bcoca commented on lib/ansible/playbook/play.py:
```diff
@@ -97,19 +98,41 @@ def __init__(self):
+                        raise AnsibleParserError("Hosts list cannot contain values of 'None'. Please check your playbook")
+                    elif not is_string(entry):
+                        raise AnsibleParserError(err_msg.format(host=entry))
+                    elif not is_string(entry):
```
these 2 lines are dupe of both above

bcoca commented:
note that if play had `name` you would get the following error before this PR
```
  File "/home/bcoca/work/ansible/lib/ansible/vars/manager.py", line 496, in _get_magic_variables
    _hosts_all = [h.name for h in self._inventory.get_hosts(pattern=pattern, ignore_restrictions=True)]
  File "/home/bcoca/work/ansible/lib/ansible/inventory/manager.py", line 391, in get_hosts
    if pattern_hash not in self._hosts_patterns_cache:
TypeError: unhashable type: 'AnsibleSequence'
```

bcoca commented on lib/ansible/playbook/play.py:
```diff
@@ -97,19 +98,41 @@ def __init__(self):
+                        raise AnsibleParserError(err_msg.format(host=entry))
+                    elif not is_string(entry):
+                        raise AnsibleParserError(err_msg.format(host=entry))
+
```
remove 116 and 117 as dupes

samdoran committed:
Remove duplicate code

s-hertel commented on lib/ansible/playbook/play.py:
```diff
@@ -97,19 +98,39 @@ def __init__(self):
+                for entry in value:
+                    if entry is None:
+                        raise AnsibleParserError("Hosts list cannot contain values of 'None'. Please check your playbook")
+                    elif not is_string(entry):
```
I had to look up is_string. It's True for vaults, which aren't supported for hosts. The error for vaulted hosts is pretty clear at least though:

```
ERROR! the field 'hosts' should be a list of (<class 'str'>,), but the item 'host1' is a <class 'ansible.parsing.yaml.objects.AnsibleVaultEncryptedUnicode'>

The error appears to be in '/home/liveuser/ansible/test_hosts.yml': line 2, column 3, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

---
- name: test hosts
  ^ here
```

samdoran replied:
Vaulted hosts 🤯. I should add a test for that. Not something I thought about.

samdoran replied:
I'll switch to an `issequence()` check to error with `AnsibleVaultEncryptedUnicode` objects.

s-hertel reviewed (approved):
Wondered about is_string vs isinstance(..., string_types), but this is a big improvement

samdoran committed:
Move error message inline

It's not shared anymore so there is no need to store it in a variable.

samdoran committed:
Use isinstance() rather than is_string()

Since is_string() considers AnsibleVaultEncryptedUnicode objects strings and those
are not valid for hosts entries, check for string types directly.
